File: backend/socket.ts
import { randomUUID as uuid } from "crypto";
import type { Player } from "./types.js";
import { GAME_CONFIG, WEAPON_CYCLE } from "./config.js";
import { allPlayers, persistentGame } from "./state.js";
import { wss } from "./server.js";
import {
  broadcast,
  debouncedBroadcastOnlineList,
  serializePlayers,
} from "./utils.js";
import { shoot, reloadWeapon, performDash, addPlayerToGame, removePlayerFromGame } from "./game.js";
import { WebSocket } from "ws";
import { getLeaderboard, getPlayerStats, getUserByToken } from "./database.js";
import { serialize, deserialize } from "./protocol.js";

export function setupSocket() {
  const HEARTBEAT_INTERVAL = 15_000;

  const heartbeat = setInterval(() => {
    wss.clients.forEach((ws) => {
      const ext = ws as WebSocket & { isAlive?: boolean };
      if (ext.isAlive === false) {
        ext.terminate();
        return;
      }
      ext.isAlive = false;
      ext.ping();
    });
  }, HEARTBEAT_INTERVAL);

  wss.on("close", () => clearInterval(heartbeat));

  wss.on("connection", (ws) => {
    let player: Player | null = null;

    const ext = ws as WebSocket & { isAlive?: boolean };
    ext.isAlive = true;
    ws.on("pong", () => { ext.isAlive = true; });

    if (wss.clients.size > 200) {
      ws.close(1013, "Server is full");
      return;
    }

    // Send online player count
    ws.send(serialize({ type: "onlineCount", count: allPlayers.size }));

    ws.on("message", (msg) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      let data: any;
      try {
        data = deserialize(msg as Buffer);
      } catch {
        return;
      }
      if (!data || typeof data.type !== "string") return;

      /* ================= ANTI-CHEAT: RATE LIMITING ================= */
      if (player) {
        const now = Date.now();
        if (now - player.msgWindowStart > 1000) {
          player.msgCount = 0;
          player.msgWindowStart = now;
        }
        player.msgCount++;
        if (player.msgCount > 120) {
          player.violations++;
          if (player.violations >= 5) {
            console.log(`🚫 Kicked ${player.username} for message flooding`);
            ws.close(4001, "Rate limit exceeded");
            return;
          }
          return;
        }
      }

      /* ================= JOIN (enter game immediately) ================= */

      if (data.type === "join") {
        if (player !== null) return;

        let trimmed: string | undefined;

        if (data.token && typeof data.token === "string") {
          const user = getUserByToken(data.token);
          if (user) trimmed = user.username;
          else {
            ws.send(serialize({ type: "error", message: "Session expired." }));
            return;
          }
        }

        if (!trimmed) {
          if (!data.username || typeof data.username !== "string") return;
          trimmed = data.username.trim();
        }

        const username = trimmed as string;

        if (
          username.length < GAME_CONFIG.USERNAME_MIN_LENGTH ||
          username.length > GAME_CONFIG.USERNAME_MAX_LENGTH ||
          !GAME_CONFIG.USERNAME_PATTERN.test(username)
        ) {
          ws.send(serialize({
            type: "error",
            message: `Username must be ${GAME_CONFIG.USERNAME_MIN_LENGTH}-${GAME_CONFIG.USERNAME_MAX_LENGTH} characters (letters, numbers, _).`,
          }));
          return;
        }

        const usernameTaken = Array.from(allPlayers.values()).some(
          (p) => p.username.toLowerCase() === username.toLowerCase()
        );
        if (usernameTaken) {
          ws.send(serialize({ type: "error", message: "Name already in use." }));
          return;
        }

        if (!persistentGame) {
          ws.send(serialize({ type: "error", message: "Server starting up..." }));
          return;
        }

        player = {
          id: uuid(),
          username,
          ws,
          team: 0,
          x: 0,
          y: 0,
          hp: GAME_CONFIG.PLAYER_HP,
          shots: GAME_CONFIG.SHOTS_PER_MAGAZINE,
          reloading: false,
          lastShotTime: 0,
          keys: { w: false, a: false, s: false, d: false },
          lastProcessedInput: 0,
          kills: 0,
          deaths: 0,
          score: 0,
          ready: true,
          aimAngle: 0,
          weapon: "machinegun",
          skin: Number(data.skin) || 0,
          speedBoostUntil: 0,
          minigunUntil: 0,
          killStreak: 0,
          lastKilledBy: "",
          shieldUntil: 0,
          invisibleUntil: 0,
          regenUntil: 0,
          lastRegenTick: 0,
          armor: 0,
          dashCooldownUntil: 0,
          dashUntil: 0,
          dashDirX: 0,
          dashDirY: 0,
          msgCount: 0,
          msgWindowStart: Date.now(),
          violations: 0,
          lastWeaponSwitch: 0,
        };

        allPlayers.set(player.id, {
          id: player.id,
          username: player.username,
          status: "in-game",
          ws: player.ws,
        });
        debouncedBroadcastOnlineList();

        addPlayerToGame(player, persistentGame);

        // Send full game state to joining player
        ws.send(serialize({
          type: "gameJoined",
          playerId: player.id,
          username,
          players: serializePlayers(persistentGame),
          obstacles: persistentGame.obstacles,
          orbs: persistentGame.orbs.map((o) => [o.id, o.x, o.y]),
          arenaWidth: GAME_CONFIG.ARENA_WIDTH,
          arenaHeight: GAME_CONFIG.ARENA_HEIGHT,
          maxHp: GAME_CONFIG.PLAYER_HP,
        }));

        // Notify others
        broadcast(persistentGame, {
          type: "playerJoined",
          username: player.username,
          playerId: player.id,
        });

        // Broadcast updated count
        const countMsg = serialize({ type: "onlineCount", count: allPlayers.size });
        wss.clients.forEach((c) => {
          try { if (c.readyState === WebSocket.OPEN) c.send(countMsg); } catch { /* ignore */ }
        });

        return;
      }

      if (data.type === "getLeaderboard") {
        const stats = getLeaderboard(10);
        ws.send(serialize({ type: "leaderboard", stats }));
        return;
      }

      if (!player) return;

      /* ================= IN-GAME ACTIONS ================= */

      const game = persistentGame;
      if (!game) return;
      if (!game.players.some((p) => p.id === player!.id)) return;

      if (data.type === "chat") {
        const chatMsg = String(data.message || "").trim().slice(0, 100);
        if (!chatMsg) return;
        broadcast(game, { type: "chatMessage", username: player.username, message: chatMsg });
        return;
      }

      if (data.type === "reload") { reloadWeapon(player); return; }
      if (data.type === "dash") { performDash(player); return; }

      if (data.type === "getMyStats") {
        const stats = getPlayerStats(player.username);
        ws.send(serialize({ type: "myStats", stats }));
        return;
      }

      if (data.type === "selectSkin") {
        const skinIndex = Number(data.skin);
        if (Number.isInteger(skinIndex) && skinIndex >= 0 && skinIndex <= 7) {
          player.skin = skinIndex;
        }
        return;
      }

      const key = data.key as "w" | "a" | "s" | "d";
      const type = data.type as "keydown" | "keyup" | "shoot" | "aim" | "switchWeapon";

      if (type === "keydown" && (key === "w" || key === "a" || key === "s" || key === "d")) {
        player.keys[key] = true;
        player.lastProcessedInput = data.sequence || 0;
      }
      if (type === "keyup" && (key === "w" || key === "a" || key === "s" || key === "d")) {
        player.keys[key] = false;
        player.lastProcessedInput = data.sequence || 0;
      }

      if (type === "shoot") {
        let dirX = Number(data.dirX) || 0;
        let dirY = Number(data.dirY) || -1;
        const mag = Math.sqrt(dirX * dirX + dirY * dirY);
        if (mag > 0.001) { dirX /= mag; dirY /= mag; }
        else { dirX = 0; dirY = -1; }
        shoot(player, game, dirX, dirY);
      }

      if (type === "aim") {
        const angle = Number(data.aimAngle);
        if (Number.isFinite(angle)) player.aimAngle = angle;
      }

      if (type === "switchWeapon") {
        if (player.weapon === "minigun") return;
        const now = Date.now();
        if (now - player.lastWeaponSwitch < 250) return;
        player.lastWeaponSwitch = now;
        if (data.weapon && WEAPON_CYCLE.includes(data.weapon)) {
          player.weapon = data.weapon;
        } else {
          const currentIndex = WEAPON_CYCLE.indexOf(player.weapon);
          player.weapon = WEAPON_CYCLE[(currentIndex + 1) % WEAPON_CYCLE.length];
        }
        const maxAmmo =
          player.weapon === "shotgun" ? GAME_CONFIG.SHOTGUN_AMMO :
          player.weapon === "sniper" ? GAME_CONFIG.SNIPER_AMMO :
          GAME_CONFIG.SHOTS_PER_MAGAZINE;
        if (player.shots > maxAmmo) player.shots = maxAmmo;
      }
    });

    ws.on("error", (err) => {
      console.error(`⚠️ WebSocket error for ${player?.username || "unknown"}:`, err.message);
    });

    ws.on("close", () => {
      if (!player) return;
      console.log(`👋 ${player.username} disconnected`);

      allPlayers.delete(player.id);
      debouncedBroadcastOnlineList();

      if (persistentGame) {
        removePlayerFromGame(player.id, persistentGame);
        broadcast(persistentGame, {
          type: "playerDisconnected",
          username: player.username,
          playerId: player.id,
        });
      }

      const countMsg = serialize({ type: "onlineCount", count: allPlayers.size });
      wss.clients.forEach((c) => {
        try { if (c.readyState === WebSocket.OPEN) c.send(countMsg); } catch { /* ignore */ }
      });
    });
  });
}

File: backend-go/socket.go
package main

import (
	"fmt"
	"math"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 4096,
	CheckOrigin: func(r *http.Request) bool {
		return true // Allow all origins
	},
	EnableCompression: true,
}

// startHeartbeat pings all connected players every 15s.
// Writes go through each player's ConnMu to avoid concurrent-write panics.
func startHeartbeat() {
	const interval = 15 * time.Second
	ticker := time.NewTicker(interval)
	go func() {
		for range ticker.C {
			allPlayersMu.RLock()
			tps := make([]*TrackedPlayer, 0, len(allPlayers))
			for _, tp := range allPlayers {
				tps = append(tps, tp)
			}
			allPlayersMu.RUnlock()

			for _, tp := range tps {
				tp.ConnMu.Lock()
				if tp.Conn != nil {
					_ = tp.Conn.SetWriteDeadline(time.Now().Add(5 * time.Second))
					_ = tp.Conn.WriteMessage(websocket.PingMessage, nil)
				}
				tp.ConnMu.Unlock()
			}
		}
	}()
}

func handleWebSocket(w http.ResponseWriter, r *http.Request) {
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		fmt.Printf("⚠️ WebSocket upgrade error: %v\n", err)
		return
	}

	// Check capacity
	allPlayersMu.RLock()
	currentCount := len(allPlayers)
	allPlayersMu.RUnlock()
	if currentCount >= 200 {
		conn.WriteMessage(websocket.CloseMessage,
			websocket.FormatCloseMessage(websocket.CloseTryAgainLater, "Server is full"))
		conn.Close()
		return
	}

	// Setup pong handler
	conn.SetReadLimit(1024)
	conn.SetReadDeadline(time.Now().Add(30 * time.Second))
	conn.SetPongHandler(func(string) error {
		conn.SetReadDeadline(time.Now().Add(30 * time.Second))
		return nil
	})

	// Send initial online count
	allPlayersMu.RLock()
	count := len(allPlayers)
	allPlayersMu.RUnlock()
	initMsg, _ := Serialize(map[string]interface{}{
		"type":  "onlineCount",
		"count": count,
	})
	conn.WriteMessage(websocket.BinaryMessage, initMsg)

	var player *Player
	var playerMu sync.Mutex // protects local player pointer

	// Cleanup on disconnect
	defer func() {
		playerMu.Lock()
		p := player
		playerMu.Unlock()

		if p != nil {
			fmt.Printf("👋 %s disconnected\n", p.Username)

			allPlayersMu.Lock()
			delete(allPlayers, p.ID)
			allPlayersMu.Unlock()

			debouncedBroadcastOnlineList()

			game := getPersistentGame()
			if game != nil {
				game.mu.Lock()
				removePlayerFromGame(p.ID, game)
				game.mu.Unlock()

				broadcast(game, map[string]interface{}{
					"type":     "playerDisconnected",
					"username": p.Username,
					"playerId": p.ID,
				})
			}

			broadcastOnlineList()
		}

		conn.Close()
	}()

	// Read loop
	for {
		_, msg, err := conn.ReadMessage()
		if err != nil {
			break
		}

		m, err := Deserialize(msg)
		if err != nil {
			continue
		}
		msgType := safeString(m, "type", "")
		if msgType == "" {
			continue
		}

		/* ================= RATE LIMITING ================= */
		playerMu.Lock()
		p := player
		playerMu.Unlock()

		if p != nil {
			now := unixMs()
			if now-p.MsgWindowStart > 1000 {
				p.MsgCount = 0
				p.MsgWindowStart = now
			}
			p.MsgCount++
			if p.MsgCount > 300 {
				p.Violations++
				if p.Violations >= 10 {
					fmt.Printf("🚫 Kicked %s for message flooding\n", p.Username)
					p.ConnMu.Lock()
					conn.WriteMessage(websocket.CloseMessage,
						websocket.FormatCloseMessage(4001, "Rate limit exceeded"))
					p.ConnMu.Unlock()
					break
				}
				continue
			}
		}

		/* ================= JOIN ================= */
		if msgType == "join" {
			playerMu.Lock()
			if player != nil {
				playerMu.Unlock()
				continue
			}
			playerMu.Unlock()

			username := strings.TrimSpace(safeString(m, "username", ""))
			if len(username) < GameConfig.UsernameMinLen ||
				len(username) > GameConfig.UsernameMaxLen ||
				!GameConfig.UsernamePattern.MatchString(username) {
				errMsg, _ := Serialize(map[string]interface{}{
					"type":    "error",
					"message": fmt.Sprintf("Username must be %d-%d characters (letters, numbers, _).", GameConfig.UsernameMinLen, GameConfig.UsernameMaxLen),
				})
				conn.WriteMessage(websocket.BinaryMessage, errMsg)
				continue
			}

			// Check duplicate username
			allPlayersMu.RLock()
			taken := false
			for _, tp := range allPlayers {
				if strings.EqualFold(tp.Username, username) {
					taken = true
					break
				}
			}
			allPlayersMu.RUnlock()

			if taken {
				errMsg, _ := Serialize(map[string]interface{}{
					"type":    "error",
					"message": "Name already in use.",
				})
				conn.WriteMessage(websocket.BinaryMessage, errMsg)
				continue
			}

			game := getPersistentGame()
			if game == nil {
				errMsg, _ := Serialize(map[string]interface{}{
					"type":    "error",
					"message": "Server starting up...",
				})
				conn.WriteMessage(websocket.BinaryMessage, errMsg)
				continue
			}

			skinVal := safeInt(m, "skin", 0)
			now := unixMs()

			game.mu.Lock()

			newPlayer := &Player{
				ID:       uuid.New().String(),
				ShortID:  game.NextShortID,
				Username: username,
				Conn:     conn,
				Team:     0,
				X:        0,
				Y:        0,
				HP:       GameConfig.PlayerHP,
				Shots:    GameConfig.ShotsPerMag,
				Reloading: false,
				LastShotTime:     0,
				Keys:             Keys{},
				LastProcessedInput: 0,
				Kills:  0,
				Deaths: 0,
				Score:  0,
				Ready:  true,
				AimAngle: 0,
				Weapon:   WeaponMachinegun,
				Skin:     skinVal,

				SpeedBoostUntil: 0,
				MinigunUntil:    0,
				KillStreak:      0,
				LastKilledBy:    "",
				ShieldUntil:     0,
				InvisibleUntil:  0,
				RegenUntil:      0,
				LastRegenTick:   0,
				Armor:           0,

				DashCooldownUntil: 0,
				DashUntil:         0,
				DashDirX:          0,
				DashDirY:          0,

				WaitingForRespawn: false,
				MsgCount:          0,
				MsgWindowStart:    now,
				Violations:        0,
				LastWeaponSwitch:  0,
			}
			game.NextShortID++

			// Add to allPlayers
			allPlayersMu.Lock()
			allPlayers[newPlayer.ID] = &TrackedPlayer{
				ID:       newPlayer.ID,
				Username: newPlayer.Username,
				Status:   "in-game",
				Conn:     conn,
				ConnMu:   &newPlayer.ConnMu,
			}
			allPlayersMu.Unlock()

			debouncedBroadcastOnlineList()

			addPlayerToGame(newPlayer, game)

			// Build shortIdMap
			shortIdMap := make(map[string]interface{})
			for _, gp := range game.Players {
				shortIdMap[fmt.Sprintf("%d", gp.ShortID)] = map[string]interface{}{
					"id":       gp.ID,
					"username": gp.Username,
				}
			}

			elapsed := now - game.MatchStartTime
			roundDurationMs := int64(GameConfig.RoundDuration) * 1000
			timerRemaining := int((roundDurationMs - elapsed) / 1000)
			if timerRemaining < 0 {
				timerRemaining = 0
			}

			joinedMsg, _ := Serialize(map[string]interface{}{
				"type":        "gameJoined",
				"playerId":    newPlayer.ID,
				"shortId":     newPlayer.ShortID,
				"username":    username,
				"players":     serializePlayers(game),
				"obstacles":   serializeObstacles(game.Obstacles),
				"orbs":        serializeOrbs(game.Orbs),
				"arenaWidth":  GameConfig.ArenaWidth,
				"arenaHeight": GameConfig.ArenaHeight,
				"maxHp":       GameConfig.PlayerHP,
				"shortIdMap":  shortIdMap,
				"timerRemaining": timerRemaining,
			})

			game.mu.Unlock()

			// Send to the joining player
			newPlayer.ConnMu.Lock()
			conn.WriteMessage(websocket.BinaryMessage, joinedMsg)
			newPlayer.ConnMu.Unlock()

			// Notify others
			broadcast(game, map[string]interface{}{
				"type":     "playerJoined",
				"username": newPlayer.Username,
				"playerId": newPlayer.ID,
				"shortId":  newPlayer.ShortID,
			})

			broadcastOnlineList()

			playerMu.Lock()
			player = newPlayer
			playerMu.Unlock()

			continue
		}

		/* ================= REQUIRE PLAYER ================= */
		if p == nil {
			continue
		}

		game := getPersistentGame()
		if game == nil {
			continue
		}

		/* ================= IN-GAME ACTIONS ================= */

		switch msgType {
		case "chat":
			chatMsg := strings.TrimSpace(safeString(m, "message", ""))
			if len(chatMsg) > 100 {
				chatMsg = chatMsg[:100]
			}
			if chatMsg == "" {
				continue
			}
			broadcast(game, map[string]interface{}{
				"type":     "chatMessage",
				"username": p.Username,
				"message":  chatMsg,
			})

		case "reload":
			game.mu.Lock()
			reloadWeapon(p)
			game.mu.Unlock()

		case "dash":
			game.mu.Lock()
			performDash(p)
			game.mu.Unlock()

		case "requestRespawn":
			game.mu.Lock()
			requestRespawn(p, game)
			game.mu.Unlock()

		case "selectSkin":
			skinIndex := safeInt(m, "skin", -1)
			if skinIndex >= 0 && skinIndex <= 7 {
				p.Skin = skinIndex
			}

		case "keydown":
			key := safeString(m, "key", "")
			seq := safeInt(m, "sequence", 0)
			switch key {
			case "w":
				p.Keys.W = true
			case "a":
				p.Keys.A = true
			case "s":
				p.Keys.S = true
			case "d":
				p.Keys.D = true
			}
			p.LastProcessedInput = seq

		case "keyup":
			key := safeString(m, "key", "")
			seq := safeInt(m, "sequence", 0)
			switch key {
			case "w":
				p.Keys.W = false
			case "a":
				p.Keys.A = false
			case "s":
				p.Keys.S = false
			case "d":
				p.Keys.D = false
			}
			p.LastProcessedInput = seq

		case "shoot":
			dirX := safeFloat(m, "dirX", 0)
			dirY := safeFloat(m, "dirY", -1)
			mag := math.Sqrt(dirX*dirX + dirY*dirY)
			if mag > 0.001 {
				dirX /= mag
				dirY /= mag
			} else {
				dirX = 0
				dirY = -1
			}
			game.mu.Lock()
			shoot(p, game, dirX, dirY)
			game.mu.Unlock()

		case "aim":
			angle := safeFloat(m, "aimAngle", 0)
			if !math.IsNaN(angle) && !math.IsInf(angle, 0) {
				p.AimAngle = angle
			}

		case "switchWeapon":
			now := unixMs()
			if now-p.LastWeaponSwitch < 250 {
				continue
			}
			p.LastWeaponSwitch = now

			requestedWeapon := WeaponType(safeString(m, "weapon", ""))
			if requestedWeapon != "" {
				// Check if the requested weapon is in the cycle
				found := false
				for _, w := range WeaponCycle {
					if w == requestedWeapon {
						found = true
						break
					}
				}
				if found {
					p.Weapon = requestedWeapon
				}
			} else {
				// Cycle to next weapon
				currentIdx := -1
				for i, w := range WeaponCycle {
					if w == p.Weapon {
						currentIdx = i
						break
					}
				}
				if currentIdx >= 0 {
					p.Weapon = WeaponCycle[(currentIdx+1)%len(WeaponCycle)]
				} else {
					p.Weapon = WeaponCycle[0]
				}
			}

			// Clamp ammo to new weapon's max
			var maxAmmo int
			switch p.Weapon {
			case WeaponShotgun:
				maxAmmo = GameConfig.ShotgunAmmo
			case WeaponSniper:
				maxAmmo = GameConfig.SniperAmmo
			default:
				maxAmmo = GameConfig.ShotsPerMag
			}
			if p.Shots > maxAmmo {
				p.Shots = maxAmmo
			}
		}
	}
}


